src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,347 @@
-          return js.error("WorkerRpc is unimplemented");
+  // Named intercept is enabled, this means we won't default to legacy behavior.
+  // The return value of the function is a promise that resolves once the remote returns the result
+  // of the RPC call.
+  return jsg::JsValue(js.wrapReturningFunction(js.v8Context(), [this, methodName=kj::str(name)]
+      (jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+        auto& ioContext = IoContext::current();
+        // Wait for the RPC to resolve and then process the result.
+        return js.wrapSimplePromise(ioContext.awaitIo(js, sendWorkerRpc(js, methodName, args),
+            [](jsg::Lock& js, capnp::Response<rpc::CallResults> result) -> jsg::Value {
+          return handleRpcResponse(js, kj::mv(result));
+        }));
+      }
+  ));
+}
+
+// The capability that lets us call remote methods over RPC.
+// The client capability is dropped after each callRequest().
+class JsRpcTargetImpl final : public rpc::JsRpcTarget::Server {
+public:
+  JsRpcTargetImpl(
+      kj::Own<kj::PromiseFulfiller<void>> callFulfiller,
+      IoContext& ctx,
+      kj::Maybe<kj::StringPtr> entrypointName)
+      : callFulfiller(kj::mv(callFulfiller)), ctx(ctx), entrypointName(entrypointName) {}
+
+  // Handles the delivery of JS RPC method calls.
+  kj::Promise<void> call(CallContext callContext) override {
+    auto methodName = kj::heapString(callContext.getParams().getMethodName());
+    auto serializedArgs = callContext.getParams().getSerializedArgs().getV8Serialized().asBytes();
+
+    // Try to execute the requested method.
+    return ctx.run(
+        [this, methodName=kj::mv(methodName), serializedArgs=kj::mv(serializedArgs),
+          entrypointName=entrypointName, callContext] (Worker::Lock& lock) mutable ->
+            kj::Promise<void> {
+
+      jsg::Lock& js = lock;
+      if (!FeatureFlags::get(js).getJsRpc()) {
+        // JS RPC is not enabled on the server side, we cannot call any methods.
+        auto rpcDisabled =
+            "The receiving Worker does not allow its methods to be called over RPC."_kjc;
+        JSG_FAIL_REQUIRE(Error, rpcDisabled);
+      }
+
+      // We will try to get the function, if we can't we'll throw an error to the client.
+      auto fn = tryGetFn(lock, entrypointName, ctx, methodName);
+
+      // We have a function, so let's call it and serialize the result for RPC.
+      // We account for both the JS method throwing and our serialization attempt failing.
+      // If the function returns a promise we will wait for the promise to finish so we can
+      // serialize the result.
+      // TODO(now): Do we need to use awaitJs here? `call()` has to return a kj::Promise, so maybe
+      // we can return some other promise that resolves when this jsg::Promise does?
+      return ctx.awaitJs(js, callFuncAndSerialize(js, fn, serializedArgs).then(js,
+          [callContext](jsg::Lock& js, SerializedResult result) mutable {
```

Comment:
Since you're calling `.then()` on a JS promise, the lambda you provide here is a JS heap object. But, `callContext` is a KJ I/O object. So, you need an IoOwn here. The easiest thing to do is wrap the whole lambda in `ctx.addFunctor`

```suggestion
          ctx.addFunctor([callContext](jsg::Lock& js, SerializedResult result) mutable {
```